### PR TITLE
Fix integration tests running with VFS retention enabled

### PR DESCRIPTION
**Pull request by wolfs**
We want that all our integration tests pass with VFS retention enabled.

### CODE REVIEW

**Comment by wolfs** [comment]
Things we need to fix for VFS retention to work:
- when cleaning up empty source files, notify the global VFS
- when cleaning up stale output files, notify the global VFS
- `GeneratedSingletonFileTree` needs to notify the VFS when it creates the file - or replace the type by a task for generating the Manifest
- Only watch parents which are directories. For missing files, the parent can also be a file, and we would currently try to watch it.

There are more problems, which I would fix later:
- fix watching symlinks, aka the target of a symlink in the input/output of a task changes.
- Detect changes while the build is running.

Finally, there is some flakiness in `IncrementalJavaAvoidanceAgainstClassDirIntegrationSpec`, which causes changes to the annotation processor sources to go undetected. I still don't know what is going on there.

**Comment by wolfs** [comment]
Another problem: We currently don't inform the VFS, that the previous output is cleaned in `CleanupOutputsStep`. That isn't a problem when the previous output is in the same location, but it is a problem when the previous output is somewhere else and the task has been reconfigured.

In `BuildSrcPluginIntegrationTest` the test uses `Project.delete()` at configuration time to remove some outputs. That also goes undetected and we try to watch the now non-exisiting directory. We should probably fix this later by watching files during the build as well.

**Comment by wolfs** [comment]
There also seems to be a problem on macOS: When the parent directory is removed, then no events are received for a watched subdirectory. For example, when watching `build/classes`, and `build` is removed from the command line, we do not get any information about that. That causes some relocation tests to fail, e.g. `Antlr2RelocationIntegrationTest`.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CacheTaskArchiveErrorIntegrationTest.groovy [hunk 1]
```diff
@@ -223,8 +219,8 @@ class CacheTaskArchiveErrorIntegrationTest extends AbstractIntegrationSpec {
         errorOutput.contains("Couldn't read content of file '${link}'")
     }
 
-    private TestFile cleanBuildDir() {
-        file("build").deleteDir()
+    private void cleanBuildDir() {
+        run(":clean")
     }
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CacheTaskArchiveErrorIntegrationTest.groovy]: I wonder if this is a good way to fix the tests. Shouldn't we detect the `build` directory being removed too?

Also, the method name should be something different I think. Or perhaps it should be inlined.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedPathSensitivityIntegrationTest.groovy [hunk 2]
```diff
@@ -25,6 +25,10 @@ class CachedPathSensitivityIntegrationTest extends AbstractPathSensitivityIntegr
     def setup() {
         buildFile << """
             task clean {
+                def cleanTask = delegate
+                tasks.all { otherTask ->
+                    cleanTask.destroyables.register(otherTask.outputs.files)
+                }
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedPathSensitivityIntegrationTest.groovy]: Why is this needed?

**Comment by wolfs** [comment]
I found this Rust library which offers an OS-independent API for file watching: https://github.com/notify-rs/notify

We could steal some things from there.

**Comment by wolfs** [comment]
The failures are now understood, and I created the follow-up issues:
- Fixes for the executor and some individual tests: https://github.com/gradle/gradle/pull/11848
- Support long paths on Windows: https://github.com/gradle/gradle/issues/11832
- Investigate failing play test: https://github.com/gradle/gradle/issues/11849
- File change on macOS not recognized when whole watched directory is removed: https://github.com/gradle/gradle/issues/11833
- Re-evaluate tests again when all the above is fixed: https://github.com/gradle/gradle/issues/11850

I am closing this PR for now.
